File: android/app/src/main/java/com/anyaicliremote/app/ui/PendingUserEchoTracker.kt
package com.anyaicliremote.app.ui

internal class PendingUserEchoTracker {
    private var expectedText: String? = null
    private var consumedLength: Int = 0

    fun begin(text: String) {
        expectedText = text
        consumedLength = 0
    }

    fun consume(chunk: String): Boolean {
        val expected = expectedText ?: return false
        if (chunk.isEmpty()) return true
        val remaining = expected.substring(consumedLength)
        var advance = when {
            chunk == expected -> expected.length - consumedLength
            remaining.startsWith(chunk) -> chunk.length
            expected.startsWith(chunk) -> chunk.length - consumedLength
            else -> overlapAdvance(expected, chunk)
        }
        if (advance < 0) {
            clear()
            return false
        }
        consumedLength = (consumedLength + advance).coerceAtMost(expected.length)
        if (consumedLength == expected.length) clear()
        return true
    }

    fun clear() {
        expectedText = null
        consumedLength = 0
    }

    private fun overlapAdvance(expected: String, chunk: String): Int {
        val consumed = expected.substring(0, consumedLength)
        for (overlap in minOf(consumed.length, chunk.length) downTo 1) {
            if (consumed.endsWith(chunk.substring(0, overlap)) &&
                expected.startsWith(chunk.substring(overlap), consumedLength)
            ) return chunk.length - overlap
        }
        return -1
    }
}

File: ios/AnyAICLIRemote/Core/PendingUserEchoTracker.swift
import Foundation

final class PendingUserEchoTracker {
    private var expectedText: String?
    private var consumedLength = 0

    func begin(text: String) { expectedText = text; consumedLength = 0 }

    func consume(chunk: String) -> Bool {
        guard let expected = expectedText else { return false }
        if chunk.isEmpty { return true }
        let expectedCharacters = Array(expected)
        let chunkCharacters = Array(chunk)
        let consumed = min(consumedLength, expectedCharacters.count)
        let remaining = String(expectedCharacters.dropFirst(consumed))
        let advance: Int
        if chunk == expected { advance = expectedCharacters.count - consumed }
        else if remaining.hasPrefix(chunk) { advance = chunkCharacters.count }
        else if expected.hasPrefix(chunk) { advance = chunkCharacters.count - consumed }
        else { advance = overlapAdvance(expectedCharacters, chunkCharacters, consumed: consumed) }
        guard advance >= 0 else { clear(); return false }
        consumedLength = min(consumed + advance, expectedCharacters.count)
        if consumedLength == expectedCharacters.count { clear() }
        return true
    }

    func clear() { expectedText = nil; consumedLength = 0 }

    private func overlapAdvance(_ expected: [Character], _ chunk: [Character], consumed: Int) -> Int {
        let consumedCharacters = Array(expected.prefix(consumed))
        guard !chunk.isEmpty else { return 0 }
        for overlap in stride(from: min(consumedCharacters.count, chunk.count), through: 1, by: -1) {
            if Array(consumedCharacters.suffix(overlap)) == Array(chunk.prefix(overlap)) &&
                Array(expected.dropFirst(consumed).prefix(chunk.count - overlap)) == Array(chunk.dropFirst(overlap)) {
                return chunk.count - overlap
            }
        }
        return -1
    }
}
